### PR TITLE
Flask+Redis Nulecule app

### LINKED ISSUE #3: Create Nulecule example for Flask + Redis app
Reference - https://github.com/dharmit/flask_redis_nulecule_app

### PR DESCRIPTION
**Pull request by dharmit**
Fixes #3

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,2 @@
+.vagrant/
+Vagrantfile
```
Comment (navidshaikh) [inline .gitignore]: This could be part of separate PR.

File: flask-redis-centos7-atomicapp/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,9 @@
+FROM projectatomic/atomicapp
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/Dockerfile]: You might want to add the tag for the image.

File: flask-redis-centos7-atomicapp/Nulecule [hunk 3]
```diff
@@ -0,0 +1,31 @@
+---
+specversion: 0.0.2
+id: flask_redis_nulecule
+
+metadata:
+    name: Flask Redis Nulecule App
+    appversion: 0.0.1
+    description: Atomic app for deploying Flask + Redis page visit counter
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/Nulecule]: "Atomic App of a simple light weight Flask and Redis app. This Flask app counts the number of time page is visited and stores the counter in Redis." sounds simple to me.

File: flask-redis-centos7-atomicapp/Nulecule [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+graph:
+    - name: redis
+      params: 
+        - name: redisimage
+          description: Redis image
+          default: dharmit/redis
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/Nulecule]: Kindly provide the image tag.

File: flask-redis-centos7-atomicapp/Nulecule [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+      artifacts:
+        docker:
+          - file://artifacts/docker/redis
+        kubernetes:
+          - file://artifacts/kubernetes/00-service-redis.yaml
+          - file://artifacts/kubernetes/01-pod-redis.yaml
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/Nulecule]: The artifacts filenames do not necessarily start with numbers stating the order in which objects to be created. They can be simple as "redis-service.yaml" , "redis-rc.yaml" and so on.

File: flask-redis-centos7-atomicapp/Nulecule [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    - name: flask
+      params: 
+        - name: flaskimage
+          description: Flask image
+          default: dharmit/flask
+      artifacts:
+        docker:
+          - file://artifacts/docker/flask
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/Nulecule:34]: Specify the image tag.

File: flask-redis-centos7-atomicapp/README.md [hunk 7]
```diff
@@ -0,0 +1,34 @@
+###Overview
+
+This is an atomic application based on Nulecule specification. This is a 2-tier
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/README.md]: *This is an atomicapp [..]

File: flask-redis-centos7-atomicapp/README.md [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+application based on [flask](flask.pocoo.org) and [redis](redis.io). Every time
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/README.md]: *Flask

File: flask-redis-centos7-atomicapp/README.md [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+###Usage
+
+To run this app, workstation needs to have atomic command. Use below command to
+start the app:
+
+    $ atomic run dharmit/flask_redis
+
+To check the app, open [http://localhost:5000](http://localhost:5000) in the
+web browser.
+
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/README.md:30]: Check out https://github.com/projectatomic/nulecule-library/blob/master/gitlab-centos7-atomicapp/README.md and provide the multiple options to spin the app.

File: flask-redis-centos7-atomicapp/artifacts/docker/flask [hunk 10]
```diff
@@ -0,0 +1 @@
+docker run -d -p 5000:5000 --link redis:redis --name flask $flaskimage
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/docker/flask]: the host port in `-p` option should not be hard coded, use env vars.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/00-service-redis.yaml [hunk 11]
```diff
@@ -0,0 +1,12 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: redis
+  labels:
+    name: redis
+spec:
+  ports:
+  - port: 6379
+    targetPort: 6379
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/00-service-redis.yaml]: `targetPort` does not necessarily be a number, it could be a string as well, which identifies the port. Infact, it should be a word defining the port, like `redis-port`. The `targetPort` defined in redis pod manifest and `targetPort` in redis service manifest should match.

Check https://github.com/projectatomic/nulecule-library/blob/master/gitlab-centos7-atomicapp/artifacts/kubernetes/redis-service.json#L10 for reference.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/01-pod-redis.yaml [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    app: redis
+spec:
+  containers:
+  - name: redis
+    image: dharmit/redis
+    command:
+    - "redis-server"
+    ports:
+    - containerPort: 6379
+      hostPort: 6379
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/01-pod-redis.yaml]: In associated `docker` provider artifact, we did not use the host Port, do we need it? 
Even if we need it, do not hard code the `hostPort` value, always use the env vars.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/01-pod-redis.yaml [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+    volumeMounts:
+    - mountPath: /redis
+      name: redis-store
+  volumes:
+  - name: redis-store
+    hostPath:
+      path: /opt/redis
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/01-pod-redis.yaml]: again use the env vars, do not hard code the values that are of hosts system.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/02-service-flask.yaml [hunk 14]
```diff
@@ -0,0 +1,11 @@
+apiVersion: v1
+kind: Service
+metadata:
+  name: flask
+  labels:
+    name: flask
+spec:
+  ports: 
+  - port: 5000
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/02-service-flask.yaml]: you can use something like `nodePort` field here and specify the port on the node (yes using env vars) where the service will be accessible.

Check https://github.com/projectatomic/nulecule-library/blob/master/gitlab-centos7-atomicapp/artifacts/kubernetes/gitlab-http-service.json

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+  name: flask
+  labels:
+    name: flask
+    app: flask
+spec:
+  containers:
+  - name: flask
+    image: dharmit/flask
+    ports:
+    - containerPort: 5000
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml]: Use name `name` field here and reference accordingly in flask service manifest by `targetPort` field.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml [hunk 16]
```diff
@@ -0,0 +1,13 @@
+apiVersion: v1
+kind: Pod
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml]: Transform this `Pod` manifest into replicationController.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/01-pod-redis.yaml [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+kind: Pod
+metadata:
+  name: redis
+  labels:
+    name: redis
+    app: redis
+spec:
+  containers:
+  - name: redis
+    image: dharmit/redis
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/01-pod-redis.yaml]: do not hard code image name in manifests, use env vars.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml [hunk 18]
```diff
@@ -0,0 +1,13 @@
...
+metadata:
+  name: flask
+  labels:
+    name: flask
+    app: flask
+spec:
+  containers:
+  - name: flask
+    image: dharmit/flask
```
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml]: do not hard code image names, use env vars.

File: .gitignore (hunk 1 above)
Comment (dharmit) [inline .gitignore]: Removed.

File: flask-redis-centos7-atomicapp/Dockerfile (hunk 2 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/Dockerfile]: Added `0.1.12`

File: flask-redis-centos7-atomicapp/Nulecule (hunk 3 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/Nulecule]: Modified.

File: flask-redis-centos7-atomicapp/Nulecule (hunk 4 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/Nulecule]: Added `latest`

File: flask-redis-centos7-atomicapp/Nulecule (hunk 5 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/Nulecule]: Done.

File: flask-redis-centos7-atomicapp/Nulecule (hunk 6 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/Nulecule:34]: Added `latest`

File: flask-redis-centos7-atomicapp/README.md (hunk 7 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/README.md]: Done.

File: flask-redis-centos7-atomicapp/README.md (hunk 8 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/README.md]: Done

File: flask-redis-centos7-atomicapp/README.md (hunk 9 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/README.md:30]: Done.

File: flask-redis-centos7-atomicapp/artifacts/docker/flask (hunk 10 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/artifacts/docker/flask]: Done.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/00-service-redis.yaml (hunk 11 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/00-service-redis.yaml]: Done.

File: flask-redis-centos7-atomicapp/artifacts/docker/flask (hunk 10 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/artifacts/docker/flask]: Done.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/01-pod-redis.yaml (hunk 13 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/01-pod-redis.yaml]: Done

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/02-service-flask.yaml (hunk 14 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/02-service-flask.yaml]: Done.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml (hunk 15 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml]: You mean `name` field here and `targetPort` field in the service manifest, right?

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml (hunk 16 above)
Comment (dharmit) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml]: Done.

File: flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml (hunk 15 above)
Comment (navidshaikh) [inline flask-redis-centos7-atomicapp/artifacts/kubernetes/03-pod-flask.yaml]: Yeah, updated my comment :)

**Comment by navidshaikh** [comment]
After updating to `docker-1.8.2-7.el7.centos.x86_64` in ADB, the app spins.
+1 from my side.
